PlayerInput 개발 및 InputSystem 로직 개선

JuicyPark opened the pull request:
### 1. BaseController와 InputManager를 연결할 PlayerInput 제작
**코드**
InputManager.cs : 각 키들의 상태값을 저장하고 있음
PlayerInput.cs : InputManager로부터 각 키들의 상태를 불러옴

### 2. Input System 로직 변경
**코드**
InputSystemEditor.cs : Input System 에디터 코드

**바뀐점**
기존 : ScriptableObject에서 키 적용시 Unity 내부 InputManager를 변경하는 방식
개선 : InputSystem이라는 에디터를 이용해서 Unity InputManager가 아닌 별도의 InputSetting 오브젝트를 생성하는 방식으로 바꿈

**바꾼이유** 
키 변경을 할경우 기존 방식은 Unity InputManager를 런타임 도중에 수정해야하는데 그런 방식보다는 별개의 InputSetting을 만들고 키 변경이 일어난경우 만들어둔 InputSetting을 수정하는게 더 나을것이라 생각하여서

### 3. 개선된 InputSystem 사용법
![image](https://user-images.githubusercontent.com/31693348/128598665-f66cf455-e9c7-4682-87ee-89850bf41123.png)
1. 상단의 InputSystem/Set Input System 버튼을 누르면 우측의 Editor가 활성화가 됨
2. 사용할 키 명단들을 정의
3. Apply했으면 앞으로 사용시에 다음과 같이 사용할 수 있는 키들의 리스트를 볼 수 있게 됨
![image](https://user-images.githubusercontent.com/31693348/128598712-27cb76c2-3ecc-4720-979f-5450e065ef06.png)

JuicyPark committed:
feat : BaseController와 InputManager를 연결할 PlayerInput 제작

JuicyPark committed:
refactor : Input System 로직 변경

Parkhyunseo commented on Union/Assets/InputSystem/Editor/InputSystemEditor.cs:
```diff
@@ -0,0 +1,120 @@
+        private SerializedObject _serializedObject;
+        private GUIStyle _titleStyle = new GUIStyle();
+
+        private string _directory = "Assets/Resources/InputSystem";
```
디렉토리 경로는 변경과 확인이 용이하게 Serialize 받는게 좋을 것 같습니다.

JuicyPark replied:
![image](https://user-images.githubusercontent.com/31693348/128634660-9dd2038f-4b46-498e-8a9d-475e70a15500.png)

좋습니다 😁
이제 경로와 파일이름을 직접 지정하고 Refresh 할 수 있습니다.

Parkhyunseo commented on Union/Assets/InputSystem/Editor/InputSystemEditor.cs:
```diff
@@ -0,0 +1,120 @@
+        private string _settingName = "InputSetting.asset";
+
+        [MenuItem("InputSystem/Set Input system")]
+        static void open()
```
Open이 되어야 할 것 같습니다.

JuicyPark replied:
저 함수는 유니티 EditorWindow 내장함수라 open으로 사용했습니다

Parkhyunseo commented on Union/Assets/InputSystem/InputManager.cs:
```diff
@@ -25,63 +25,84 @@ public static InputManager Instance
+            _inputSetting = Resources.Load<InputSetting>(_settingPath);
+            if (_inputSetting == null)
+            {
+                Debug.LogError("Input System� ������� �ʾҽ�ϴ�.");
```
텍스트가 Input SystemÀ» ¼³Á¤ÇÏÁö ¾Ê¾Ò½À´Ï´Ù. 이렇게 나오는데 정상인가요??

JuicyPark replied:
예전에 회사에서 봤던 UTF-8 인코딩관련 에러였는데 왠지 그부분인것 같네요 수정해야겠습니다

JuicyPark replied:
가 아니라 소스트리 설정문제였군요 해결완료

Parkhyunseo commented on Union/Assets/InputSystem/InputManager.cs:
```diff
@@ -25,63 +25,84 @@ public static InputManager Instance
-        private Dictionary<ButtonName, ButtonKey> _buttonKeys = new Dictionary<ButtonName, ButtonKey>();
+        private Dictionary<KeyName, KeyInput> _keyInputs = new Dictionary<KeyName, KeyInput>();
+        private Dictionary<MouseName, MouseInput> _mouseInputs = new Dictionary<MouseName, MouseInput>();
+        private string _settingPath = "InputSystem/InputSetting";
```
InputSystemEditor의 settingPath값과 참조가 아닌 값으로 묶여있어서 잠재적 버그 위험성이 있습니다.
참조 형식으로 수정이 필요해 보입니다.

JuicyPark replied:
굉장히 좋은 의견입니다.

안그래도 Editor클래스(InputSystemEditor.cs)에서는 일반클래스(InputManager.cs)로 뭔가를 전달할수없어서 어떻게 해야 InputSetting.asset의 경로를 참조할수 있을지 고민을 많이하다가 저런식으로 주먹구구식으로 넘긴건데, 고민끝에 해결방법을 찾았습니다.

InputSystemEditor.cs에서 Enum을 자동생성할때 해당 InputSetting의 경로를 담아서 전역변수로 만드는 방법으로 해결했습니다. ( InputSystemEditor.cs 126줄, InputName.cs를 참고)

Parkhyunseo commented on Union/Assets/InputSystem/InputSetting.cs:
```diff
@@ -0,0 +1,31 @@
+
+    public class InputSetting : ScriptableObject
+    {
+        public List<KeyButton> keyButtons;
```
사용자 입력은 Serialize로 받고 외부 참조는 property로 받는 구조는 어떨까요? 사소한거긴 하지만 용도의 분리가 되어서 저는 습관처럼 하고는 합니다.

Parkhyunseo reviewed (changes requested):
고생하셨습니다. 리뷰 반영 부탁드리고 질문 있으시면 남겨주세요!

JuicyPark committed:
refactor : InputManager가 받는 InputSetting Path를 참조형식으로 변경

JuicyPark committed:
refactor : public 변수 프로퍼티형식으로 변경

JuicyPark committed:
chore : 오타수정

JuicyPark committed:
chore : 디폴트 접근제한자 수정

GyutaeLee committed:
- feat : 전투 로직 제작
1. Player / Enemy 클래스 추가 및 테스트용 GamePlayer 클래스 제거
2. 테스트용 공격/피격 : player / enemy 부딪히면 서로 hp가 감소하도록 개발
3. HealthPoint 클래스 HUD hp bar 추가

JuicyPark commented on Union/Assets/Scripts/Game/BattleField.cs:
```diff
@@ -7,7 +7,7 @@ public class BattleField : Singleton<BattleField>
         private static class Constants
         {
-            public const int InitialEnemyCount = 10;
+            public const int InitialEnemyCount = 1;
```
BattleField 클래스에 적 카운트를 상수클래스 Constants로 감싸서 상수로 만든 이유가있을까요?

GyutaeLee replied:
테스트용 상수에는 "TEST_" 프리픽스를 붙여서 명확하게 표기하도록 하겠습니다. (보다 깔끔하고 좋은 방법이 있으면 추천 부탁드립니다)

JuicyPark commented on Union/Assets/Scripts/Game/GameLogic.cs:
```diff
@@ -6,11 +6,10 @@ public class GameLogic : Singleton<GameLogic>
         private static class Constants
         {
-            public const int BattleFieldEnemyCount = 10;
+            public const int BattleFieldEnemyCount = 1;
```
위 BattleField클래스에서 EnemyCount와 차이점이 무엇인가요? 따로선언한 이유가 무엇인지 모르겠습니다

GyutaeLee replied:
사용하지 않는 변수라서 제거했습니다.

JuicyPark commented on Union/Assets/Scripts/Game/GameLogic.cs:
```diff
@@ -21,57 +20,59 @@ private void Awake()
         }
 
-        private void CreateIGameState(GameState gameState)
+        private void CreateIGameState(GameStates gameStates)
```
저번주에 말씀하셧던 부분이군요.
저라면 여기서 추가적으로 이렇게 개선해볼것같습니다.

새로운 상태가 될때마다 CreateGameState라는 함수에서 스위치문을 돌면서 새로운 객체를 new해서 넣어주는것보다는 GameLogic.cs의 멤버변수로  Dictionary< GameStates(enum), GameState(class object) > 로 만들어놓고 Start()에서 초기화할때  해당 딕셔너리의 값을 채워줄것 같습니다.

enum과 class를 매칭시킬땐 Dictionary를 이용하면 깔끔하게 할 수 있습니다.

GyutaeLee replied:
Dictionary 구조로 수정했습니다.

JuicyPark commented on Union/Assets/Scripts/Game/GameLogic.cs:
```diff
@@ -6,11 +6,10 @@ public class GameLogic : Singleton<GameLogic>
 
-        private IGameState _iGameState;
-        private GameState _gameState;
+        private GameState _GameState;
```
프로퍼티가 아닌데 파스칼 케이스입니다

GyutaeLee replied:
카멜 표기법으로 변경했습니다.

JuicyPark commented on Union/Assets/Scripts/Game/GameLogic.cs:
```diff
@@ -6,11 +6,10 @@ public class GameLogic : Singleton<GameLogic>
-        private GameState _gameState;
+        private GameState _GameState;
 
         public GameTime gameTime;
```
퍼블릭 변수방식보다는
에디터에서 수정할 목적이었다면[SerializeField]를 사용하거나
외부에서 사용할 목적이었다면 프로퍼티를 이용해야할것 같습니다.

GyutaeLee replied:
프로퍼티로 수정했습니다.

JuicyPark commented on Union/Assets/Scripts/Stat/HealthPoint.cs:
```diff
@@ -0,0 +1,36 @@
+    public class HealthPoint : Stat
+    {
+        private GameObject _headUpDisplayCanavs;
+        private Image _headUpDispalyHealthPointBar;
```
HealthPoint 클래스가 _headUpDisplayCanavs, _headUpDispalyHealthPointBar를 들고있기보다는
HealthPoint는 딱 Stat으로써의 값만을 책임져야할것 같은 느낌이 더듭니다.

화면에 표시를 해주는 View 책임을 갖는 HeadUpHealthPoint.cs(이름막지음) 클래스를 만들어서
그 클래스가 HealthPoint의 값을 바라보는게 더 나을것 같은 생각이 듭니다.

GyutaeLee replied:
Stat과 HUD 클래스를 분리하고, component로 붙여서 사용할 수 있도록 했습니다.

JuicyPark commented on Union/Assets/Scripts/Stat/Stat.cs:
```diff
@@ -3,27 +3,45 @@
     public class Stat
     {
         protected int amount;
+        protected int maxAmount;
```
protected도 _maxAmount로 하는게 맞을것 같슴다

GyutaeLee replied:
protected도 언더바 추가하는거로 합의!

JuicyPark commented on Union/Assets/Scripts/UI/UIPlayerStatus.cs:
```diff
@@ -0,0 +1,43 @@
+
+        private void UpdateHealthPointUI()
+        {
+            this._healthPointBar.fillAmount = Mathf.Lerp(this._healthPointBar.fillAmount, (float)this._player.UnitStat.healthPoint.Get() / this._player.UnitStat.healthPoint.GetMax(), Time.deltaTime * 10.0f);
```
10.0f 라는 매직넘버보다는 멤버변수로 선언해서 쓰는건 어떨까요

GyutaeLee replied:
테스트 숫자도 변수로 수정했습니다.

추가로 논의하고 싶은 점이 생겼습니다.
기존에 상수면 Constants 클래스를 이너 클래스로 만드는 규칙을 합의했었는데요.
특정 함수 내부에서만 사용하는 함수는 굳이 Constants 클래스에 넣기보다는, 함수 내부에서 상수를 써도 무방할거 같다는 생각이 들었습니다. 다음 회의때 얘기해보면 좋을듯 합니다.

JuicyPark commented on Union/Assets/Scripts/Unit/Enemy.cs:
```diff
@@ -0,0 +1,75 @@
+            this._enemyState?.Enter();
+        }
+
+        private void CreateEnemyState(EnemyStates enemyStates)
```
여기도 위에서 GameLogic.cs 말씀드린부분과 같습니다

GyutaeLee replied:
수정했습니다.

JuicyPark reviewed (commented):
고생하셨습니다

Parkhyunseo commented on Union/Assets/Scripts/UI/UIPlayerStatus.cs:
```diff
@@ -0,0 +1,43 @@
+        private void UpdateHealthPointUI()
+        {
+            this._healthPointBar.fillAmount = Mathf.Lerp(this._healthPointBar.fillAmount, (float)this._player.UnitStat.healthPoint.Get() / this._player.UnitStat.healthPoint.GetMax(), Time.deltaTime * 10.0f);
+            this._healthPointText.text = "HP : " + this._player.UnitStat.healthPoint.Get().ToString();
```
이러면 메모리 릭이 발생할 수 있으니
StringBuilder + Cache를 해야합니다!

Parkhyunseo replied:
https://docs.unity3d.com/kr/2019.4/Manual/UnderstandingAutomaticMemoryManagement.html

GyutaeLee replied:
StringBuilder 로 수정했습니다.

GyutaeLee committed:
Revert "- feat : 전투 로직 제작"

This reverts commit db99f3a06035549144f09b571e5ee0288519714d.

JuicyPark committed:
fix : InputSetting 적용시 자동으로 AssetDatabase 저장되도록 수정

JuicyPark committed:
refactor : public 변수들 프로퍼티 방식으로 변경

JuicyPark committed:
refactor : 마우스 키 인풋을 따로두지않고 하나의 방식으로 통합

JuicyPark committed:
chore : Unity 기본 InputManager 디폴트로 변경

Parkhyunseo commented:
여기에 규태형 추가 코멘트 작성할게요

근데 거의 준수형이 다 작성해줘서 거의 Request Change보다 Comment형식으로 적을게요

개인적으로 C#에서 Enum 사용은 Java의 Enum과 달리 사용을 최소화 하기를 바랍니다! 분기를 유발하다보니 객체지향적으로 설계하는데 어려움을 줄 때가 많아서요. 그래서 캐릭터 상태, 몬스터 상태, 게임 상태 이런 것들이 필요하기 전까지는 놔두고 필요하다고 생각될 때 추가했으면 합니다! 

그리고 만약 상태가 그래프 구조가 아니고 단방향이나 리스트 형태라면 더더욱 Enum을 안 써도 된다고 생각합니다!  만약 GameLogic이 상태의 전이가 방향이 여러 갈래이거나 역방향 이동이 있는 그래프 구조라면 FSM처럼 Transition(전이) 또는 Trigger(조건)을 만들어서 상태, 전이 를 추상화 해서 처리하는게 좋아보입니다.

JuicyPark committed:
feat : 마우스의 GetAxis값 추가

JuicyPark commented on Union/Assets/Scripts/PlayerInput.cs:
```diff
@@ -21,6 +21,6 @@ public PlayerInput()
         Jump = InputSystem.InputManager.Instance.GetKey(KeyName.Jump);
         Attack = InputSystem.InputManager.Instance.GetKey(KeyName.Attack);
-        MousePosition = InputSystem.InputManager.Instance.MousePosition;
+        MouseInput = InputSystem.InputManager.Instance.MouseInput;
```
PlayerInput에서 MouseInput을 받아와서
MouseInput.cs
{
    Vector2 Position;
    float X;
    float Y;
}
를 사용할 수 있도록 수정했습니다

JuicyPark commented on Union/Assets/InputSystem/MouseInput.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace InputSystem
+{
+    public class MouseInput
```
현서님이 이전 작업했던 PlayerTwoAxisAction 처럼 따로 클래스를 파지않고 Mouse에대해서만 TwoAxis를 만든이유는

1. 현재 InputSystem은 Unity 내장 InputManager를 이용하는 방식이 아니기때문입니다. Input.GetAxis("Mouse X") 이런 코드처럼 Unity 내장 InputManager에 명시된 키값을 직접 불러오는 방식을 사용하지 않기때문
2. 마우스를 제외한곳에서 TwoAxis형식으로 키입력을 받는 구간이 적거나 없을것이라 판단해서, 특별한 케이스인 마우스에만 다음과 같이 Input.GetAxis("Mouse X"); 같이 예외적으로 사용하기로 생각하였기때문

JuicyPark committed:
chore : MouseInput 선언 누락된 부분 추가

Parkhyunseo reviewed (approved):
확인했습니다!
